### ISSUE TITLE
[HOLD for payment 2021-11-09] Reduce the haptic feedback vibration to a short bip 

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#4355]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!

Upwork posting: https://www.upwork.com/jobs/~0105c031baf3dfd3df
___

## Action Performed:
1. Navigate to a conversation in mobile
2. Press and hold on a message
 
## Expected Result:
There should be a short blip of vibration. 

## Actual Result:
Current vibration is longer that it should be. 

To achieve the desired result, we should:
1. Update our app to use react-native-haptic-feedback
2. Fork the library, and update the forked lib to have the same vibration as slack does when selecting messages
2. Once the PR is merged into our codebase, create a PR upstream so that the functionality can be merged into react-native-haptic-feedback

## Workaround:
N/A

## Platform:

Where is this issue occurring?

- iOS
- Android

**Version Number:** 1.0.82-0

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

**Expensify/Expensify Issue URL:**

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

___ 
From @quinthar https://expensify.slack.com/archives/C01GTK53T8Q/p1627669368027700

> I love the new haptic feedback when you long tap on mobile. However, it uses a kind of strange vibration.  Can we match the same vibration that slack uses with long tap? It's just a incredibly short blip.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @pecanoro (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @Jag96 (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by Jag96** [comment]
It looks like there's some more context on this issue in [this slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1627677384030500?thread_ts=1627669368.027700&cid=C01GTK53T8Q). cc @rushatgabhane since you had some good thoughts here.

**Comment by rushatgabhane** [comment]
@Jag96 I'm not able to find any npm package which matches the short blip of Slack.

However, I can get the exact haptic feedback as Slack's using native code for Android.

Here are our options for adding native code.
1. Directly into New Expensify using [NativeModules](https://reactnative.dev/docs/native-modules-intro).
2. Create a new npm package.

Which one do you prefer?

Alternatively, we could settle for a shorter duration of the **strange** vibration using [rn-haptic-feedback](https://www.npmjs.com/package/react-native-haptic-feedback).

**Comment by Jag96** [comment]
We've used NativeModules before [here](https://github.com/Expensify/App/blob/a946115eb9e8ac76e29974036af1ca8cf7c70e41/src/libs/Environment/betaChecker/index.ios.js#L10) so I think that would be fine, but I'd reply to the thread in [slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1627669368027700) (and send to the main channel as well by ticking `Also send to expensify-open-source`) just to see if anybody has any opinions.

And just to confirm, using NativeModules would you just override the index.android.js functionality and continue using the current lib for iOS?

**Comment by rushatgabhane** [comment]
> And just to confirm, using NativeModules would you just override the index.android.js functionality and continue using the current lib for iOS?

No. I won't reusing the current lib at all.
NativeModules will be used both for iOS and Android.

**Comment by rushatgabhane** [comment]
I forgot to mention, simulators don't support haptics and I don't own an iPhone.
I'm gonna use apple's [haptics docs](https://developer.apple.com/design/human-interface-guidelines/ios/user-interaction/haptics/) for reference if that's fine.

**Comment by rushatgabhane** [comment]
Update from [slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1627669368027700)

We decided to add `blip()` method to [Vibration](https://reactnative.dev/docs/vibration) API by creating a PR for React Native.

I won't be available for a few weeks. So I won't be taking this issue.
But here are things that might help with implementation.

Android: `Vibrate(createPredefined(VibrationEffect.EFFECT_CLICK))`
iOS: [Selection](https://developer.apple.com/design/human-interface-guidelines/ios/user-interaction/haptics/)

**Comment by quinthar** [comment]
Tiny thing: I vote naming the API call "bip" instead of "blip".  When you
say it out loud, "bip" is closer to the actual vibrate we're going for.

**Comment by MelvinBot** [comment]
@Jag96 Uh oh! This issue is overdue by 2 days. Don't forget to update your issues!

**Comment by arpitdeveloper** [comment]
In this case we add the code  
	{<TouchableOpacity
             
             onLongPress={() => Vibration.vibrate(1 * ONE_SECOND_IN_MS)}
         >
	....
	</TouchableOpacity>}
in InvertedFlatList in "/src/pages/home/report/ReportActionsView.js"

**Comment by chira37** [comment]
change src\components\PressableWithSecondaryInteraction\index.native.js


```

import {Vibration} from 'react-native';

<Pressable
...
        onLongPress={(e) => {
            e.preventDefault();
            Vibration.vibrate(20);
            props.onSecondaryInteraction(e);   
        }}
...
    >
        {props.children}
    </Pressable> ````

in src\components\PressableWithSecondaryInteraction

it will reduce time from 50 (with haptics library) millisecond to 20 millisecond

**Comment by Jag96** [comment]
@arpitdeveloper @chira37 please see the description of this issue for the desired implementation:

To achieve the desired result, we should:
1. Update the Vibration lib in our [react native fork](https://github.com/Expensify/react-native/tree/master/Libraries/Vibration) to add a new function `bip` that will use the shortest, strongest vibration possible (see [slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1627669368027700))
2. Once the PR is merged, create a PR upstream so that the functionality can be merged into React Native's [Vibration library](https://reactnative.dev/docs/vibration)

**Comment by silverfoxgit** [comment]
We can reduce the time of vibration. 
This is possible in React Native.

**Comment by mallenexpensify** [comment]
Doubled price to $500 in Upwork.

**Comment by chira37** [comment]
> @arpitdeveloper @chira37 please see the description of this issue for the desired implementation:
> 
> To achieve the desired result, we should:
> 
> 1. Update the Vibration lib in our [react native fork](https://github.com/Expensify/react-native/tree/master/Libraries/Vibration) to add a new function `bip` that will use the shortest, strongest vibration possible (see [slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1627669368027700))
> 2. Once the PR is merged, create a PR upstream so that the functionality can be merged into React Native's [Vibration library](https://reactnative.dev/docs/vibration)

I have added the bip() function to the expensify react native fork PR was created

**Comment by Jag96** [comment]
@chira37 please ensure you've read through the slack thread, and please refrain from making PRs until your proposal has been accepted.

From the [slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1627669368027700), the current `vibrate` function does not allow for any customization on iOS, and the vibration on Android is not the vibration we want. The `bip` function should make use of the native libraries mentioned [in the slack thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1627961579109900?thread_ts=1627669368.027700&cid=C01GTK53T8Q). If you have any questions, feel free to ask!

**Comment by MelvinBot** [comment]
@Jag96 it looks like no one is assigned to work on this job.
Please double the price or add a comment stating why the job isn't being doubled.

**Comment by Jag96** [comment]
Doubled the price: https://www.upwork.com/jobs/~0105c031baf3dfd3df

**Comment by MTN718** [comment]
Yes agree. so first we need to try use react native-haptic-feedback soft method if it's not working, we can customize library

**Comment by Jag96** [comment]
@roryabraham that's great context I wasn't aware of, I agree that getting rid of expo-haptics is a good idea.

> Yes agree. so first we need to try use react native-haptic-feedback soft method if it's not working, we can customize library

@MTN718 are you able to confirm the fix you are proposing will fix the issue? The accepted proposal for this job should be able to clearly state what technical changes are required to fix the issue, not what steps will be taken if the proposed fix doesn't work. It sounds like you are not sure if the `react native-haptic-feedback soft method` will work here, and you haven't explained what technical changes would have to be made if the methods for the trigger function don't have the effect. If I'm misunderstanding something please let me know!

One question I have @roryabraham, why would it be better to fork `react-native-haptic-feedback` and create a PR there instead of just updating our existing react-native fork to add the new `bip` method to the Vibrate library? In both scenarios, we will no longer be using `expo-haptics`, so the only difference would be that we would have to fork another library which I would think we'd want to avoid. Is there something I'm missing?

**Comment by roryabraham** [comment]
@Jag96 There are a few reasons I think it would be better to fork `react-native-haptic-feedback` than to attempt to implement `bip` directly in `react-native`:

- Getting a PR merged and included in a React Native release is going to be much harder than doing so in `react-native-haptic-feedback`. In general I think we should only update our fork for bug fixes we feel have a strong chance of being accepted by Facebook, because otherwise our fork and the upstream will diverge, making it harder to upgrade RN versions in the future.
- From what I can tell, `react-native-haptic-feedback` is not a direct replacement for the `Vibrate` library, nor does it use the `Vibrate` library internally. It provides its own native-module implementation with greater capabilities and customization options than the `Vibrate` library. For example, I don't think that the `Vibrate` library offers any customizations for vibration strength.
- `react-native-haptic-feedback` simplifies the use of a number of different haptic feedback modes, which could enable us to enable other forms of haptic feedback in the future – something I have no doubt we'll want to do.
- I'm very skeptical that RN would ever merge such an improvement in `Vibrate`, for a number of reasons. First, it's a rather specific use-case. Because of the current differences between `react-native-haptic-feedback` and `Vibrate`, it sounds like it might require a fair deal of re-writing in their native modules, and I have a feeling that they would only be interested in a more holistic set of more general-purpose upgrades to enable greater customization of haptic feedback. Even then, there's a big push at Facebook for React Native's core to remain lean, so they _might_ even take out the `Vibrate` library in preference of a community package, like they did for other deprecated packages such as `AsyncStorage`. Meanwhile, I think our chances of getting the `bip` option we want with the desired effect merged into `react-native-haptic-feedback` are pretty good.

**Comment by MTN718** [comment]
@Jag96 react-native-haptic-feedback library cover all vibration type and it has added native codes to customize vibration format. so using it we can create vibrate format as we want. I'm sure.

**Comment by Jag96** [comment]
@roryabraham that makes sense to me, thanks for the explanation.

@MTN718 you still have not explained what technical changes are required to create the vibrate format we want, so it is still unclear to me what code changes are required to get the desired effect on both Android and iOS.

**Comment by MTN718** [comment]
![image](https://user-images.githubusercontent.com/12084368/132094704-de4edefb-e92a-4262-831b-93cb3fadba4a.png)
I will create new method on forked react-native-haptic-feedback. As you can see image, we can change duration value for android and for IOS, we will use soft.

![image](https://user-images.githubusercontent.com/12084368/132094731-8840daa4-4451-484e-a7c2-75602c6775c0.png)

**Comment by rushatgabhane** [comment]
Hi, I just got relieved of my obligation so I can take this issue if you like.

## Proposal
Looks like we're gonna be using `react-native-haptic-feedback`.

For iOS: `react-native-haptic-feedback` already has the [selection ](https://developer.apple.com/design/human-interface-guidelines/ios/user-interaction/haptics/#selection)method for iOS implemented [(source)](https://github.com/junina-de/react-native-haptic-feedback/blob/8e4d598ad9be886325316b0e2140df8df624a91f/ios/RNReactNativeHapticFeedback.m#L51
). 
So we can just `RNHapticFeedback.trigger('selection')` for 

Android: 
Predefined effects are a set of common vibration effects that should be identical, regardless of the app they come from. Works only on Android 10 and above, and falls back to a generic pattern.
[source](https://developer.android.com/reference/android/os/VibrationEffect#createPredefined(int))

We'll add an android only method `effectClick`.
`v.vibrate(createPredefined(VibrationEffect.EFFECT_CLICK))`

**Comment by rushatgabhane** [comment]
Hi, I just got relieved of my obligation so I can take this issue if you like.

## Proposal
Looks like we're gonna be using `react-native-haptic-feedback`.

### For iOS: 
`react-native-haptic-feedback` already has [selection ](https://developer.apple.com/design/human-interface-guidelines/ios/user-interaction/haptics/#selection), which is the desired method implemented. ([source](https://github.com/junina-de/react-native-haptic-feedback/blob/8e4d598ad9be886325316b0e2140df8df624a91f/ios/RNReactNativeHapticFeedback.m#L51)).
So we can just check if the platform is iOS and call `RNHapticFeedback.trigger('selection')` in [PressableWithSecondaryInteraction ](https://github.com/Expensify/App/blob/main/src/components/PressableWithSecondaryInteraction/index.native.js)component.

### Android:
Predefined effects are a set of common vibration effects that should be identical, regardless of the app they come from. Works only on Android 10 and above, and falls back to a generic pattern.
([source](https://developer.android.com/reference/android/os/VibrationEffect#createPredefined(int)))

We'll fork `react-native-haptic-feedback` and add an android only method `effectClick`.
Which would essentially call
`v.vibrate(createPredefined(VibrationEffect.EFFECT_CLICK))`

This will only work for Android 8 and above.
So we can check for it using `Build.VERSION.SDK_INT >= Build.VERSION_CODES.O`, and add a small vibration of 10ms as a fall back.

**Comment by Jag96** [comment]
Updated the description to take into account we're looking to update react-native-haptic-feedback now instead of RN directly. Working internally to get somebody to help me test the proposals on a physical Android device to confirm the vibration matches slack, thanks for bearing with me @MTN718 and @rushatgabhane

**Comment by Jag96** [comment]
After some testing, it seems that `RNHapticFeedback.trigger('selection')` gives the desired effect on iOS (the `soft` setting is not the desired effect). I also had multiple people test on physical Android devices (@Julesssss and @Beamanator) and they both confirmed that react-native-haptic-feedback's 'soft' setting in [this testing PR](https://github.com/Expensify/App/compare/main...joe-test-android-vibration?expand=1) did match Slack's vibration pattern after enabling haptic feedback as an OS feature.

@rushatgabhane are you suggesting that `Vibrate(createPredefined(VibrationEffect.EFFECT_CLICK))` is required to match the desired intensity, and that no other methods in [this list](https://www.npmjs.com/package/react-native-haptic-feedback#available-methods) match that on android? Just double-checking here after the internal testing

**Comment by MTN718** [comment]
Regarding IOS provides only 5 types of vibration so if soft is not working, we can test other types

**Comment by rushatgabhane** [comment]
> are you suggesting that Vibrate(createPredefined(VibrationEffect.EFFECT_CLICK)) is required to match the desired intensity, and that no other methods in this list match that on android? Just double-checking here after the internal testing

@Jag96 Yes! I tested [this PR](https://github.com/Expensify/App/compare/main...joe-test-android-vibration?expand=1) and it is nothing like slack's haptic feedback for long press a message.

Device: Pixel 3, Android 11 up-to-date.

I also tried all the methods from the list again, but none of them match slack.

**Comment by rushatgabhane** [comment]
> are you suggesting that Vibrate(createPredefined(VibrationEffect.EFFECT_CLICK)) is required to match the desired intensity, and that no other methods in this list match that on android? Just double-checking here after the internal testing

@Jag96 Yes! I tested [the PR you mentioned](https://github.com/Expensify/App/compare/main...joe-test-android-vibration?expand=1) (`trigger.(soft)`)and it is nothing like slack's haptic feedback for long press a message.

Device: Pixel 3, Android 11 up-to-date.

I also tried all the methods from the list again, but none of them match slack.

**Comment by rushatgabhane** [comment]
I'll link a PR which implements my proposed changes soon, I hope it'll add more clarity.

**Comment by rushatgabhane** [comment]
I'll link a PR which implements my proposed changes soon, I hope it'll add more clarity.

**Comment by rushatgabhane** [comment]
@Jag96 [here is the PR](https://github.com/rushatgabhane/Expensify-App/tree/test-haptic-android) which implements `VibrationEffect.EFFECT_CLICK`

Please test it out on Android 9+.

**Comment by Jag96** [comment]
@rushatgabhane thanks for the info and for making the example PR! @Julesssss @Beamanator could you help me out in testing @rushatgabhane's branch to confirm it matches slack on your devices as well? If you're unable to build to your device from that fork, I've also updated [my PR](https://github.com/Expensify/App/compare/main...joe-test-android-vibration?expand=1) to use @rushatgabhane's branch.

From the testing above, it sounds like `soft` can match slack on some devices, but it may not on others, so I'd like to test here to confirm on multiple devices.

**Comment by Jag96** [comment]
Still held on n6, clearing overdue label

**Comment by Jag96** [comment]
Still held on n6

**Comment by kadiealexander** [comment]
Please refer to [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1633557259458600) for updated information on the `n6 hold`, we've raised the bonus to **$250** for all issues where a **PR is created** before the N6 hold is lifted.

**Comment by rushatgabhane** [comment]
@Jag96 should I submit a PR to NewDot?

--
I've taken the liberty to add new methods to `rn-haptic-feedback` (like effectClick, effectHeavyClick etc). 
And also made the code more modular.  [PR link](https://github.com/junina-de/react-native-haptic-feedback/pull/66)

I hoped that [my PR](https://github.com/junina-de/react-native-haptic-feedback/pull/66)  to `rn-haptic-feedback` would have been merged by now.
But there has been no reply on it for 4 weeks.

Is there anything that I can do to get it merged faster?
Should I email a maintainer?

**Comment by Jag96** [comment]
@rushatgabhane this is still on n6-hold, and we'll need to create an Expensify fork of `rn-haptic-feedback` first for this where you can then submit a PR, so I think we can hold for now. If the n6-hold label comes off and the PR still hasn't been reviewed in the rn-haptic-feedback repo then we can get that repo forked by Expensify, have you create a PR on the Expensify fork, get it reviewed/tested, and then make a PR in NewDot to use that fork in our package.json.

Going to assign you the issue here, re-create the Upwork posting since it looks like it expired, and hire you for the job.

Since there hasn't been any reply for a bit I think it's fine to email the maintainer to see if they missed it, you can also try requesting a review from `mkuczera` (and tagging them in the PR description) since it looks like they have reviewed many of the more recent PRs merged to the repo.

**Comment by rushatgabhane** [comment]
Got it, thanks!

**Comment by rushatgabhane** [comment]
Hi Joe!
My [PR](https://github.com/junina-de/react-native-haptic-feedback/pull/66) for adding new haptic methods got merged into `junina-de/react-native-haptic-feedback`. 

So I created a new draft PR (5960) which uses it, and fixes this issue.

Expensify's fork of `rn-haptic-feedback` isn't needed now.

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.12-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/5960


If no regressions arise, payment will be issued on 2021-11-09. :confetti_ball:

**Comment by Jag96** [comment]
Paid, including n6 hold bonus!

**Comment by rushatgabhane** [comment]
Thanks a ton!
